File: daily_problems/2024/12/1202/personal_submission/cf362c_yawn_sean.cpp
#include <bits/stdc++.h>
// #pragma GCC optimize("O3,Ofast,unroll-loops")
// #include "atcoder/convolution"
// #include "atcoder/dsu"
// #include "atcoder/fenwicktree"
// #include "atcoder/lazysegtree"
// #include "atcoder/math"
// #include "atcoder/maxflow"
// #include "atcoder/mincostflow"
// #include "atcoder/modint"
// #include "atcoder/scc"
// #include "atcoder/segtree"
// #include "atcoder/string"
// #include "atcoder/twosat"
// using namespace __gnu_cxx;
// using namespace __gnu_pbds;
using namespace std;
auto rng = mt19937(random_device()());
auto rngl = mt19937_64(random_device()());

// Let's set a bit and flow!

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(0);
    cout.tie(0);

    int n;
    cin >> n;

    vector<int> perm(n);
    for (auto &x: perm)
        cin >> x;
    
    vector<vector<int>> pref(n + 1, vector<int>(n + 1, 0));
    for (int i = 0; i < n; i ++)
        pref[i + 1][perm[i] + 1] = 1;
    
    for (int i = 0; i <= n; i ++) {
        for (int j = 0; j < n; j ++) {
            pref[i][j + 1] += pref[i][j];
        }
    }
    
    for (int i = 0; i < n; i ++) {
        for (int j = 0; j <= n; j ++) {
            pref[i + 1][j] += pref[i][j];
        }
    }

    int ans = 0;
    for (int i = 0; i < n; i ++) {
        for (int j = 0; j < i; j ++) {
            ans += (perm[i] < perm[j]);
        }
    }

    int res = -1, cnt = 0;

    for (int i = 0; i < n; i ++) {
        for (int j = 0; j < i; j ++) {
            if (perm[i] < perm[j]) {
                int v = 2 * (pref[i][perm[j]] - pref[j + 1][perm[j]] - pref[i][perm[i] + 1] + pref[j + 1][perm[i] + 1]) + 1;
                if (v > res) {
                    res = v;
                    cnt = 1;
                }
                else if (v == res)
                    cnt ++;
            }
        }
    }

    cout << ans - res << ' ' << cnt;

    return 0;
}

File: daily_problems/2024/12/1202/personal_submission/cf362c_yawn_sean.py
# Submission link: https://codeforces.com/contest/362/submission/294280933
def main():
    n = II()
    perm = LGMI()

    ans = 0
    for i in range(n):
        for j in range(i):
            if perm[i] < perm[j]:
                ans += 1

    def f(x, y):
        return x * (n + 1) + y

    pref = [0] * ((n + 1) * (n + 1))
    for i in range(n):
        pref[f(i + 1, perm[i] + 1)] = 1

    for i in range(n + 1):
        for j in range(n):
            pref[f(i, j + 1)] += pref[f(i, j)]

    for i in range(n):
        for j in range(n + 1):
            pref[f(i + 1, j)] += pref[f(i, j)]

    res = -1
    cnt = 0

    for i in range(n):
        for j in range(i):
            if perm[i] < perm[j]:
                v = 2 * (pref[f(i, perm[j])] - pref[f(j + 1, perm[j])] - pref[f(i, perm[i] + 1)] + pref[f(j + 1, perm[i] + 1)]) + 1
                if v > res:
                    res = v
                    cnt = 1
                elif v == res:
                    cnt += 1

    print(ans - res, cnt)